### ISSUE TITLE
Charging status and percentage

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#974]
The charging status and percentage of both halves must be displayed on the OLED display.

### Charging status

There are 3 charging statuses: charging, not charging, and battery unavailable (as the UHK 80 will be available without batteries via the webshop).

`chargerStatCallback()` should be used to update the charging state. if `gpio_pin_get_dt(&chargerStatDt)` is 0, charging is in progress; otherwise, the keyboard half is not being charged.

If the battery is absent, `chargerStatCallback()` is called twice per second with `gpio_pin_get_dt(&chargerStatDt)` alternating between 0 and 1.

~The right half has a [hardware bug I'm trying to fix](https://github.com/UltimateHackingKeyboard/firmware/issues/852) that causes the `chargerStatCallback()` to get called gazillions of times per second with a 0 stat value. Let's be aware of this bug, but don't work around it, as I'll fix it on the hardware level.~

Let's keep charging (chargerEnDt) always enabled for now. Later, we can implement a battery conservation mode by disabling charging above 70 or 80 percent.

### Charging percentage

See `cmd_uhk_charger()` for the relevant code.

1. This is the raw ADC value of battery voltage:

```
        err = adc_read(adc_channel.dev, &sequence);
        int32_t val_mv = (int32_t)buf;
```

Measure and monitor this value in a thread at a reasonable interval, such as 5 or 10 seconds. Whenever it changes, make the left half send it to the right half. The right half should process the raw ADC battery measurements as follows.

2. Convert the raw ADC measurement to mV using adc_raw_to_millivolts_dt()

3. The mV value must be converted to percentage, which is a logarithmic-shaped curve. I've made several measurements with the `scripts/log-battery.mjs` script while [dis]charging the battery in `scripts/battery-stats`.

Some wise words from Martin, our electronics consultant:

> Log it out your USB port, and let it run from fully charged to discharged.
> In addition to measuring voltage, it is a good idea to periodically check the current also.
> If current is steady over the full time duration, the # of mA times # of hours should be close to the mAhr capacity of your battery.

If you want to make battery logs, I suggest charging the halves independently with a decent charger that can output at least 3A. In practice, the charging rate will not exceed 2.3A even at the beginning of charging. Alternatively, you can charge one half via the other half, but it'll be slower. A (or two) battery charging meter can come in handy, and I recommend the Fnirsi FNB58 by far. We can purchase one or two, or you can purchase, and we can reimburse you via your next invoice.

Ignore the TS (thermistor) values. We don't want to measure battery temperature via its thermistor anymore, as it wouldn't be helpful and would require additional parts. I've removed the TS ADC node from the devicetree and from the code, but it's still present in `scripts/log-battery.mjs`.

As you can see, the measured voltage slightly fluctuates. We must combat this fluctuation, as it would confuse the user if it were visible.

As for converting the battery voltage to percentage, I think we should use an array to map raw ADC values starting from raw ADC min (indexed as 0) to raw ADC max, mapped to the percentage, based on the stats. This map would differ per keyboard half, given that their batteries have different capacities.

Now that I think about it, the mV value is irrelevant to the above solution as the raw ADC value would directly map to percentages, but it's worth being aware that the raw ADC value is not mV.

### Bluetooth battery percentage reporting

The UHK should report the charging level to the host via Bluetooth so that operating systems can display it. I'm unfamiliar with the details, but Google for `bas_notify()`. Ideally, the UHK could expose the charging percentage of both halves, but I'm not sure whether Bluetooth and the OSes support it. If it's only possible to expose a single percentage value, let's expose the minimum of both.

**Comment by mondalaci** [comment]
@kareltucek When you get to this, let me know if you have questions and how you plan to combat battery voltage measurement fluctuations.

**Comment by benedekkupper** [comment]
Just to confirm, BLE battery reporting can only support a single 0-100 percent value.

It would also be possible to report the charging state to the host OS, over USB HID, there is a kind of standard for it, although I have never really used it before. It did come to my mind already, as that interface could be used as an indication that the UHK is connected to USB, in case the BLE channel is used for keyboard functionality.

**Comment by kareltucek** [comment]
> As you can see, the measured voltage slightly fluctuates. We must combat this fluctuation, as it would confuse the user if it were visible.

Super-simple ideas that come to mind:
- always take the minimum since last charing
- compute average over a window of last n measurements
- or combination of the above - take minimum of some window averages...

May reevaluate this when I see the actual data...

> I suggest charging the halves independently with a decent charger that can output at least 3A.

Any reason for that? Shouldn't it be Voltage that matters?

> A (or two) battery charging meter can come in handy

Why? Sounds to me like having log of UHK measurements over time should be all that's needed?

> scripts/log-battery.mjs

I am confused. How do I use it? (It is a good custom to include either short help when no arguments are passed, or just a short comment in the file.)

Edit: note to self: search email for "Charging percentage calculation with/without charger"

**Comment by kareltucek** [comment]
...also, where do I connect the batteries 😅?

**Comment by mondalaci** [comment]
Your fluctuation-related ideas are sound. Taking the minimum while charging and the maximum while discharging may make sense.

A 3A charger is not required, but it allows for faster charging when testing charging/discharging cycles.

A charging meter might not be useful for you, indeed. It was useful for me to test whether the electronics can fast-charge.

I've fixed the log-battery.mjs script and added a helpful header.

See the battery connectors' connections below. I don't remember which PCB version I provided you, and your prototype may feature less robust battery connectors, but their location should be the same.

![IMG20240527215405](https://github.com/UltimateHackingKeyboard/firmware-uhk80/assets/300402/6206f74c-b2a3-4717-bfe8-4caec0b6115b)

**Comment by kareltucek** [comment]
Allright, where do I get the `ATTRS{serial}=="CPA8b119D15"` etc for the udev rules?

(I am also confused as to which cables should remain connected for battery testing, and consequently whose serials those are.)

**Comment by mondalaci** [comment]
Run `dmesg` after connecting the serial adapters individually to figure out their serial numbers.

You need to connect the Tag-Connect cables, and the USB to serial adapters must be connected to the Tag-Connect adapters for logging purposes. The J-Link adapters might cause slight unintended charging, so it's better to disconnect them.

In the Zephyr shell, `uhk statlog 1` enables STAT logging, which reveals whether the battery is being charged.

**Comment by kareltucek** [comment]
That's the small red board, that connects to PCB by two jumper cables, right?

I tried that, but both return the same "0001".

**Comment by kareltucek** [comment]
> Taking the minimum while charging and the maximum while discharging may make sense.

Shouldn't it be maximum while charging and minimum while discharging?

**Comment by mondalaci** [comment]
No. I'd rather be conservative and take the minimum while charging, assuming that occasional high spikes can occur, skewing the measured value. The same logic applies to discharge, but it's reversed.

**Comment by kareltucek** [comment]
I mean, minimum (over entire charging period) is always the discharged value (measured when started charging). While the maximum (over entire discharging period) is always the fully charged battery (measured when started discharging).

Even if we limit it to a minimum/maximum over some window, what you are saying may fluctuate in a non-monotonic manner (e.g., battery level sometimes increasing while discharging).

The thing you say would make sense to me if we take maximum (over entire charging period) of minimums (over some window) while charging (and reverse for discharge).

**Comment by mondalaci** [comment]
Yes, I meant over some window, but I may be wrong. I trust your judgment, so feel free to implement the solution you think is the best. Fluctuations should be avoided, as they could disturb users.

**Comment by kareltucek** [comment]
Do you have an idea what should be the cut-off voltage? 

When cut-off voltage is reached, what should be done then? 

> Now that I think about it, the mV value is irrelevant to the above solution as the raw ADC value would directly map to percentages, but it's worth being aware that the raw ADC value is not mV.

Is it fundamentally different from voltage? It seems to me as just a different unit for the same thing.

> This map would differ per keyboard half, given that their batteries have different capacities.

Would it? If they are the same kind of battery, their discharge curves should be the same except for the time axis being more stretched out. If we are computing percentage based on voltage, then the same mapping function can be used.

**Comment by mondalaci** [comment]
I'm unsure what you mean by the cut-off voltage. If the battery is fully charged, charging stops, and STAT becomes 1. If you mean the battery conservation mode I mentioned, mobile operating systems stop charging at 80%:

![image](https://github.com/UltimateHackingKeyboard/firmware-uhk80/assets/300402/7bb106e7-9fb3-4baf-8d0e-3fa60df6051c)

(If the battery conservation mode is enabled, we could show a leaf icon in the status bar.)

I think the raw ADC and mV values are linearly related.

Both LiPo batteries are made by the same manufacturer, but their charge curves may differ due to their differing capacity. If they don't, feel free to use the same mapping table.

**Comment by kareltucek** [comment]
I mean voltage under which we should stop discharging the battery and shut down instead. 

The battery seems to have built-in protection at 3.0V, but sources recommend cut off voltage between 3.2V and 3.7V. Discharging the battery too low significantly reduces battery lifespan... 

(Yes, we should implement the upper bound too, but you have already answered the question how.)

(As for battery conservation mode, I think we should implement it, but it basically means to set both the voltage parameters (cut off voltage and maximum charge) to a narrower band.)

> In electronics, the cut-off voltage is the voltage at which a battery is considered fully discharged, beyond which further discharge could cause harm. Some electronic devices, such as cell phones, will automatically shut down when the cut-off voltage has been reached.

**Comment by mondalaci** [comment]
Great idea! I thought we should let the battery reach its built-in protection voltage, but as you correctly said, that's not optimal for battery health, and causes the LEDs to become yellowish. Based on some Google searching, I'd pick 3.2V.

**Comment by kareltucek** [comment]
Allright, now the more important question: how? Do we have a special circuit to disconnect the battery? Or should we just shut down the firmware?

**Comment by mondalaci** [comment]
The battery can't be disconnected on the hardware level, but if we enter sleep mode, the hardware should consume next to nothing. I'm unfamiliar with the related APIs, so we should Google for something like "nrf zephyr sleep".

I think there are multiple nRF sleep modes, and we should probably use the deepest sleep mode. Feel free to paste relevant URLs you find.

I think it's a matter of implementing UltimateHackingKeyboard/firmware#978 and immediately entering sleep mode if we're below the cut-off voltage.

**Comment by kareltucek** [comment]
> chargerStatCallback() should be used to update the charging state. if gpio_pin_get_dt(&chargerStatDt) is 0, charging is in progress; otherwise, the keyboard half is not being charged.

What exactly does chargerStatDt monitor and what does it depend on?

**Comment by mondalaci** [comment]
chargerStatDt monitors the STAT pin of the [BQ25303J](https://www.ti.com/lit/ds/symlink/bq25303j.pdf?ts=1717454131209) charger IC.

**Comment by mhantsch** [comment]
My left keyboard half seems to discharge quickly even if I am not using it. I got the device from László fully charged, and had it in the box for about one day. Unpacked, and left half was at 39%, right half still full. Charged it on USB (connected to right half) to 63%, then unplugged. This morning it was at 55%. Put it on USB and charged left half to 75%. Put it back in the box for maybe 10 hours, and in the evening it was at 66%. Right half is still 100%, has been all the time.

@mondalaci asked me to write my experience into this issue.

**Comment by mhantsch** [comment]
And next question: I now fully charged both halves. The OLED shows `100+ 100%` in the top right corner. why `+`? Shouldn't it be `100% 100%`?

**Comment by mondalaci** [comment]
`+` means that it's still being charged, and it should eventually transition to `%`.

**Comment by kareltucek** [comment]
> And next question: I now fully charged both halves. The OLED shows 100+ 100% in the top right corner. why +? Shouldn't it be 100% 100%? 

Battery-related functionality is not implemented yet. Consider the 100 to be a mock value.

As to battery life, yes, there may be some problem that's spinning the CPU.

**Comment by mhantsch** [comment]
I had to reset the two halves due to something else acting up, and now the battery drain is gone. Left it at 100% for a day, now still at 100%.

**Comment by rpnfan** [comment]
My left half drained super quickly in a few hours from full to empty (with lights on), the right side kept the charge much better -- more than I would expect from the battery capacity difference.

Update: the right side was on the whole night and discharged only a bit, while the left side has run down to reported 3 %. When plugging the keyboard in the right side percentage jumped quickly from something about 30 % to 50 %, which is not possible. The left side also jumped from 3 % to about 30 %, which can not be charged that quick (through the spiral cable anyways). So the percentages surely seem to be off.

**Comment by kareltucek** [comment]
Are you on fresh master?

On friday, I had pushed some changes that fixed a bug that might have caused fast discharge of left half.

**Comment by kareltucek** [comment]
> When plugging the keyboard in the right side percentage jumped quickly from something about 30 % to 50 %, which is not possible.

We know. The battery system is not implemented at all yet. The "percentage" is just a linear interpolation of Voltage measured, which of course jumps up when you connect power source.

**Comment by rpnfan** [comment]
I think that was already on the latest friday version, but honestly it is hard to keep track of which version is active, because all files are named the same. Would be much easier, when one could directly see which subversion one is using (Agent and firmware).

**Comment by kareltucek** [comment]
Does Agent not tell you current git commit hash?

**Comment by rpnfan** [comment]
No, not in the installer name of Agent, nor in the About screen. For the firmware I sometimes have seen the commit hash, but that is not displayed always.

**Comment by kareltucek** [comment]
Can this be closed?

**Comment by mondalaci** [comment]
Yes. Closing it.
